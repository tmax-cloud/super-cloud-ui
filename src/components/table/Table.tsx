import * as React from 'react';
import * as _ from 'lodash-es';
import TableHead from './TableHead';
import TableBody from './TableBody';
import useRequest from '../../apis/useRequest';
import { K8sKind, RequestType } from '../../types';

function Table(props: TableProps) {
  const { tableItems, kindObj } = props;
  const { isLoaded, data, errorMsg } = useRequest(kindObj, RequestType.LIST);
  return (
    <>
      {isLoaded && (
        <MuiTable aria-label="simple table">
          <TableHead tableItems={tableItems} />
          <TableBody items={data?.items} tableItems={tableItems} errorMsg={errorMsg as string} />
        </MuiTable>
      )}
    </>
  );
}

Table.defaultProps = {
  tableItems: [
    { name: 'name', displayTitle: 'Name', className: '' },
    { name: 'namespace', displayTitle: 'Namespace', className: '' },
    { name: 'type', displayTitle: 'Type', className: '', ref: 'spec.type' },
    { name: 'externalIP', displayTitle: 'External IP', className: '', customValue: (item: any) => (item.spec.type !== 'LoadBalancer' ? 'No External IP' : _.map(item.status.loadBalancer.ingress, (i) => i.hostname || i.ip || '-')) },
  ],
};

export interface TableProps {
  tableItems: TableItemProps[];
  kindObj: K8sKind;
}
export interface TableItemProps {
  name: string;
  displayTitle: string;
  className: string;
  ref?: string; // 키 이름이 뭔가 맘에 안듬... 좋은 게 생각안남..
  customValue?: any;
}

export default Table;
