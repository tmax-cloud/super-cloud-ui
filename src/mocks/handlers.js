import { rest } from 'msw';

export const handlers = [
  rest.get('/api/kubernetes/api/v1/services/', (req, res, ctx) => {
    return res(
      ctx.json({
        kind: 'ServiceList',
        apiVersion: 'v1',
        metadata: {
          selfLink: '/api/v1/services',
          resourceVersion: '105697845',
        },
        items: [
          {
            metadata: {
              name: 'api-gateway',
              namespace: 'api-gateway-system',
              selfLink: '/api/v1/namespaces/api-gateway-system/services/api-gateway',
              uid: '3169d866-b4a8-4b24-8064-3e2766cf9919',
              resourceVersion: '1211332',
              creationTimestamp: '2021-12-13T02:40:18Z',
              labels: {
                'app.kubernetes.io/component': 'reverse-proxy',
                'app.kubernetes.io/managed-by': 'tmax-cloud',
                'app.kubernetes.io/name': 'traefik',
                'app.kubernetes.io/part-of': 'api-gateway',
              },
              annotations: {
                'kubectl.kubernetes.io/last-applied-configuration':
                  '{"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"labels":{"app.kubernetes.io/component":"reverse-proxy","app.kubernetes.io/managed-by":"tmax-cloud","app.kubernetes.io/name":"traefik","app.kubernetes.io/part-of":"api-gateway"},"name":"api-gateway","namespace":"api-gateway-system"},"spec":{"ports":[{"name":"traefik","port":9000,"protocol":"TCP","targetPort":"traefik"},{"name":"web","port":80,"protocol":"TCP","targetPort":"web"},{"name":"websecure","port":443,"protocol":"TCP","targetPort":"websecure"},{"name":"k8s","port":6443,"protocol":"TCP","targetPort":"k8s"}],"selector":{"app":"traefik","app.kubernetes.io/component":"reverse-proxy","app.kubernetes.io/managed-by":"tmax-cloud","app.kubernetes.io/name":"traefik","app.kubernetes.io/part-of":"api-gateway"},"sessionAffinity":"None","type":"LoadBalancer"}}\n',
              },
              managedFields: [
                {
                  manager: 'controller',
                  operation: 'Update',
                  apiVersion: 'v1',
                  time: '2021-12-13T02:40:18Z',
                  fieldsType: 'FieldsV1',
                  fieldsV1: {
                    'f:status': {
                      'f:loadBalancer': {
                        'f:ingress': {},
                      },
                    },
                  },
                },
                {
                  manager: 'kubectl-client-side-apply',
                  operation: 'Update',
                  apiVersion: 'v1',
                  time: '2021-12-13T02:40:18Z',
                  fieldsType: 'FieldsV1',
                  fieldsV1: {
                    'f:metadata': {
                      'f:annotations': {
                        '.': {},
                        'f:kubectl.kubernetes.io/last-applied-configuration': {},
                      },
                      'f:labels': {
                        '.': {},
                        'f:app.kubernetes.io/component': {},
                        'f:app.kubernetes.io/managed-by': {},
                        'f:app.kubernetes.io/name': {},
                        'f:app.kubernetes.io/part-of': {},
                      },
                    },
                    'f:spec': {
                      'f:externalTrafficPolicy': {},
                      'f:ports': {
                        '.': {},
                        'k:{"port":80,"protocol":"TCP"}': {
                          '.': {},
                          'f:name': {},
                          'f:port': {},
                          'f:protocol': {},
                          'f:targetPort': {},
                        },
                        'k:{"port":443,"protocol":"TCP"}': {
                          '.': {},
                          'f:name': {},
                          'f:port': {},
                          'f:protocol': {},
                          'f:targetPort': {},
                        },
                        'k:{"port":6443,"protocol":"TCP"}': {
                          '.': {},
                          'f:name': {},
                          'f:port': {},
                          'f:protocol': {},
                          'f:targetPort': {},
                        },
                        'k:{"port":9000,"protocol":"TCP"}': {
                          '.': {},
                          'f:name': {},
                          'f:port': {},
                          'f:protocol': {},
                          'f:targetPort': {},
                        },
                      },
                      'f:selector': {
                        '.': {},
                        'f:app': {},
                        'f:app.kubernetes.io/component': {},
                        'f:app.kubernetes.io/managed-by': {},
                        'f:app.kubernetes.io/name': {},
                        'f:app.kubernetes.io/part-of': {},
                      },
                      'f:sessionAffinity': {},
                      'f:type': {},
                    },
                  },
                },
              ],
            },
            spec: {
              ports: [
                {
                  name: 'traefik',
                  protocol: 'TCP',
                  port: 9000,
                  targetPort: 'traefik',
                  nodePort: 31751,
                },
                {
                  name: 'web',
                  protocol: 'TCP',
                  port: 80,
                  targetPort: 'web',
                  nodePort: 30958,
                },
                {
                  name: 'websecure',
                  protocol: 'TCP',
                  port: 443,
                  targetPort: 'websecure',
                  nodePort: 30860,
                },
                {
                  name: 'k8s',
                  protocol: 'TCP',
                  port: 6443,
                  targetPort: 'k8s',
                  nodePort: 31407,
                },
              ],
              selector: {
                app: 'traefik',
                'app.kubernetes.io/component': 'reverse-proxy',
                'app.kubernetes.io/managed-by': 'tmax-cloud',
                'app.kubernetes.io/name': 'traefik',
                'app.kubernetes.io/part-of': 'api-gateway',
              },
              clusterIP: '10.96.225.224',
              type: 'LoadBalancer',
              sessionAffinity: 'None',
              externalTrafficPolicy: 'Cluster',
            },
            status: {
              loadBalancer: {
                ingress: [
                  {
                    ip: '192.168.9.141',
                  },
                ],
              },
            },
          },
          {
            metadata: {
              name: 'console',
              namespace: 'api-gateway-system',
              selfLink: '/api/v1/namespaces/api-gateway-system/services/console',
              uid: '9a1f0543-ed1a-4335-a604-6ca898dada0c',
              resourceVersion: '96329224',
              creationTimestamp: '2022-02-09T23:26:29Z',
              labels: {
                'app.kubernetes.io/component': 'webserver',
                'app.kubernetes.io/instance': 'console-ck-master',
                'app.kubernetes.io/managed-by': 'Helm',
                'app.kubernetes.io/name': 'console',
                'app.kubernetes.io/part-of': 'api-gateway',
                'app.kubernetes.io/version': '5.0.42.4',
                'helm.sh/chart': 'console-0.1.0',
              },
              annotations: {
                createdTime: '2022-02-10T08:26:29.288982552+09:00',
                creator: 'kubernetes-admin',
                'kubectl.kubernetes.io/last-applied-configuration':
                  '{"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"labels":{"app.kubernetes.io/component":"webserver","app.kubernetes.io/instance":"console-ck-master","app.kubernetes.io/managed-by":"Helm","app.kubernetes.io/name":"console","app.kubernetes.io/part-of":"api-gateway","app.kubernetes.io/version":"5.0.42.4","helm.sh/chart":"console-0.1.0"},"name":"console","namespace":"api-gateway-system"},"spec":{"ports":[{"name":"http","port":31303,"protocol":"TCP","targetPort":"http"}],"selector":{"app.kubernetes.io/instance":"console-ck-master","app.kubernetes.io/name":"console"},"type":"ClusterIP"}}\n',
                updatedTime: '2022-02-10T08:26:29.288982552+09:00',
                updater: 'kubernetes-admin',
              },
              managedFields: [
                {
                  manager: 'argocd-application-controller',
                  operation: 'Update',
                  apiVersion: 'v1',
                  time: '2022-02-09T23:26:29Z',
                  fieldsType: 'FieldsV1',
                  fieldsV1: {
                    'f:metadata': {
                      'f:annotations': {
                        '.': {},
                        'f:kubectl.kubernetes.io/last-applied-configuration': {},
                      },
                      'f:labels': {
                        '.': {},
                        'f:app.kubernetes.io/component': {},
                        'f:app.kubernetes.io/instance': {},
                        'f:app.kubernetes.io/managed-by': {},
                        'f:app.kubernetes.io/name': {},
                        'f:app.kubernetes.io/part-of': {},
                        'f:app.kubernetes.io/version': {},
                        'f:helm.sh/chart': {},
                      },
                    },
                    'f:spec': {
                      'f:ports': {
                        '.': {},
                        'k:{"port":31303,"protocol":"TCP"}': {
                          '.': {},
                          'f:name': {},
                          'f:port': {},
                          'f:protocol': {},
                          'f:targetPort': {},
                        },
                      },
                      'f:selector': {
                        '.': {},
                        'f:app.kubernetes.io/instance': {},
                        'f:app.kubernetes.io/name': {},
                      },
                      'f:sessionAffinity': {},
                      'f:type': {},
                    },
                  },
                },
              ],
            },
            spec: {
              ports: [
                {
                  name: 'http',
                  protocol: 'TCP',
                  port: 31303,
                  targetPort: 'http',
                },
              ],
              selector: {
                'app.kubernetes.io/instance': 'console-ck-master',
                'app.kubernetes.io/name': 'console',
              },
              clusterIP: '10.96.229.111',
              type: 'ClusterIP',
              sessionAffinity: 'None',
            },
            status: {
              loadBalancer: {},
            },
          },
        ],
      }),
    );
  }),
];
