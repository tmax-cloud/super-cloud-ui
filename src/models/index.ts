export interface K8sModelType {
  apiVersion: string;
  label: string;
  plural: string;
  abbr: string;
  namespaced: boolean;
  kind: string;
  id: string;
  labelPlural: string;
}

export const ServiceModel: K8sModelType = {
  apiVersion: 'v1',
  label: 'Service',
  plural: 'services',
  abbr: 'S',
  namespaced: true,
  kind: 'Service',
  id: 'service',
  labelPlural: 'Services',
};
