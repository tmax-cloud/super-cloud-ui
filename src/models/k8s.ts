import { K8sKind } from '../types';

export const ServiceModel: K8sKind = {
  apiVersion: 'v1',
  label: 'Service',
  plural: 'services',
  abbr: 'S',
  namespaced: true,
  kind: 'Service',
  id: 'service',
  labelPlural: 'Services',
};
