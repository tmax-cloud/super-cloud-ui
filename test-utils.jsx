// import React from 'react';
// import { render } from '@testing-library/react';
// 추후에 redux 추가하게 되면 redux Provider로 감싸주는 custom render 메서드 추가 해야됨.

export * from '@testing-library/react';
