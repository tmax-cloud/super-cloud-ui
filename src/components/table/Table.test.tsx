import { rest } from 'msw';
import * as React from 'react';
import { render, screen } from '../../../test-utils';
import { server } from '../../mocks/server';
import Table from './Table';
import { ServiceModel } from '../../models/index';

function setup() {
  return render(
    <Table
      columnDataList={[
        { name: 'name', displayTitle: 'Name', className: '' },
        { name: 'namespace', displayTitle: 'Namespace', className: '' },
      ]}
      kindObj={ServiceModel}
    />,
  );
}

    setup();
test('빈 배열 왔을 때 `해당 리소스를 찾을 수 없습니다.` 잘 뜨는지 테스트', async () => {
  server.use(
    rest.get('/api/kubernetes/api/v1/services', (req, res, ctx) => {
      return res.once(ctx.status(200), ctx.json({ items: [] }));
    }),
  );

    setup();

  const emptyBox = await screen.findByTestId('empty-box');
  expect(emptyBox).toHaveTextContent('해당 리소스를 찾을 수 없습니다.');
});

test('서버 에러 404일 때 errorMsg 잘 뜨는지 테스트', async () => {
  server.use(
    rest.get('/api/kubernetes/api/v1/services', (req, res, ctx) => {
      return res.once(ctx.status(404));
    }),
  );

    setup();

  const errorBox = await screen.findByTestId('error-box');
  expect(errorBox).toHaveTextContent('Unexpected end of JSON input');
});
