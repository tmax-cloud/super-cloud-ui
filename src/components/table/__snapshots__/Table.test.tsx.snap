// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> 스냅샷 테스트 table render 1`] = `
<table
  aria-label="simple table"
  class="MuiTable-root css-rqglhn-MuiTable-root"
>
  <thead
    class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
  >
    <tr
      class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
    >
      <th
        aria-sort="ascending"
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
        scope="col"
      >
        <span
          class="MuiButtonBase-root MuiTableSortLabel-root Mui-active css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
          role="button"
          tabindex="0"
        >
          Name
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
            data-testid="ArrowDownwardIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
            />
          </svg>
        </span>
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
        scope="col"
      >
        <span
          class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
          role="button"
          tabindex="0"
        >
          Namespace
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
            data-testid="ArrowDownwardIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
            />
          </svg>
        </span>
      </th>
    </tr>
  </thead>
  <tbody
    class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
  >
    <tr
      class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
    >
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
      >
        api-gateway
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
      >
        api-gateway-system
      </td>
    </tr>
    <tr
      class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
    >
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
      >
        console
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
      >
        api-gateway-system
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`테이블 테스트 테이블 스냅샷 테스트 1`] = `
<table
  aria-label="simple table"
  class="MuiTable-root css-rqglhn-MuiTable-root"
>
  <thead
    class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
  >
    <tr
      class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
    >
      <th
        aria-sort="ascending"
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
        scope="col"
      >
        <span
          class="MuiButtonBase-root MuiTableSortLabel-root Mui-active css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
          role="button"
          tabindex="0"
        >
          Name
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
            data-testid="ArrowDownwardIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
            />
          </svg>
        </span>
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
        scope="col"
      >
        <span
          class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
          role="button"
          tabindex="0"
        >
          Namespace
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
            data-testid="ArrowDownwardIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
            />
          </svg>
        </span>
      </th>
    </tr>
  </thead>
  <tbody
    class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
  >
    <tr
      class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
    >
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
      >
        api-gateway
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
      >
        api-gateway-system
      </td>
    </tr>
    <tr
      class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
    >
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
      >
        console
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
      >
        api-gateway-system
      </td>
    </tr>
  </tbody>
</table>
`;
